File: src/frontend/src/api/calendar.ts
export interface CalendarAttendee {
  email: string;
  display_name: string | null;
}

export interface CalendarEventTime {
  date_time: string;
  timezone: string;
}

export interface CalendarEventResponse {
  id: string;
  calendar_id: string;
  summary: string | null;
  description: string | null;
  location: string | null;
  start: CalendarEventTime | null;
  end: CalendarEventTime | null;
  attendees: CalendarAttendee[];
  html_link: string | null;
}

export interface ListEventsRequest {
  calendar_id: string;
  time_min: string;
  time_max: string | null;
  max_results: number;
}

export interface ListEventsResponse {
  events: CalendarEventResponse[];
}

export interface BusySlot {
  calendar_id: string;
  start: string;
  end: string;
}

export interface FreebusyRequest {
  time_min: string;
  time_max: string;
  timezone: string;
  calendar_ids: string[];
}

export interface FreebusyResponse {
  busy: BusySlot[];
  conflicts: BusySlot[];
  warnings: string[];
}

export interface PrepareCalendarEventRequest {
  thread_id: string | null;
  title: string;
  start_time: string;
  end_time: string | null;
  duration_minutes: number | null;
  timezone: string;
  calendar_id: string;
  organizer_email: string;
  attendees: CalendarAttendee[];
  location: string | null;
  description: string | null;
  video_conference: boolean;
  recurrence_rule: string | null;
  conflict_override: boolean;
}

export interface CalendarEventPayload {
  title: string;
  start_time: string;
  end_time: string;
  timezone: string;
  calendar_id: string;
  organizer_email: string;
  attendees: CalendarAttendee[];
  location: string | null;
  description: string | null;
  video_conference: boolean;
  recurrence_rule: string | null;
  reminders: Record<string, unknown>[];
  conflict_override: boolean;
  conflict_summary: BusySlot[];
  external_event_id?: string | null;
  calendar_action?: string | null;
}

export interface CalendarArtifactResponse {
  thread_id: string;
  work_item_id: string;
  artifact_id: string;
  version: number;
  maturity: string;
  conflicts: BusySlot[];
  warnings: string[];
  content: {
    title: string;
    start_time: string;
    end_time: string;
    timezone: string;
    calendar_id: string;
    organizer_email: string;
    attendees: CalendarAttendee[];
    location: string | null;
    description: string | null;
    video_conference: boolean;
    recurrence_rule: string | null;
    conflict_override: boolean;
    conflict_summary: BusySlot[];
  };
}

// Calendar token 仍然只在后端解密和使用。前端只发送用户显式填写的
// 时间、日历和参会人等业务字段。
export async function listCalendarEvents(payload: ListEventsRequest): Promise<ListEventsResponse> {
  const response = await fetch("/api/calendar/events/list", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Calendar events fetch failed with ${response.status}`);
  }

  return response.json() as Promise<ListEventsResponse>;
}

export async function queryCalendarFreebusy(
  payload: FreebusyRequest,
): Promise<FreebusyResponse> {
  const response = await fetch("/api/calendar/freebusy", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Calendar freebusy failed with ${response.status}`);
  }

  return response.json() as Promise<FreebusyResponse>;
}

export async function prepareCalendarEventDraft(
  payload: PrepareCalendarEventRequest,
): Promise<CalendarArtifactResponse> {
  const response = await fetch("/api/calendar/prepare/event", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Prepare calendar event failed with ${response.status}`);
  }

  return response.json() as Promise<CalendarArtifactResponse>;
}

export async function createCalendarEvent(
  payload: CalendarEventPayload,
): Promise<CalendarEventResponse> {
  const response = await fetch("/api/calendar/events", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Create calendar event failed with ${response.status}`);
  }

  return response.json() as Promise<CalendarEventResponse>;
}

export async function updateCalendarEvent(
  eventId: string,
  payload: CalendarEventPayload,
): Promise<CalendarEventResponse> {
  const response = await fetch(`/api/calendar/events/${encodeURIComponent(eventId)}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Update calendar event failed with ${response.status}`);
  }

  return response.json() as Promise<CalendarEventResponse>;
}

export async function deleteCalendarEvent(
  calendarId: string,
  eventId: string,
): Promise<void> {
  const response = await fetch(
    `/api/calendar/events/${encodeURIComponent(eventId)}?calendar_id=${encodeURIComponent(calendarId)}`,
    { method: "DELETE" },
  );

  if (!response.ok) {
    throw new Error(`Delete calendar event failed with ${response.status}`);
  }
}

File: src/backend/app/schemas/calendar.py
from typing import Any

from pydantic import BaseModel, Field


class CalendarAttendee(BaseModel):
    """日程参会人。"""

    email: str
    display_name: str | None = None


class CalendarEventTime(BaseModel):
    """Google Calendar 使用的时间结构。"""

    date_time: str
    timezone: str


class CalendarEventPayload(BaseModel):
    """本地日程 Artifact 和 Proposal payload 使用的结构。"""

    title: str
    start_time: str
    end_time: str
    timezone: str
    calendar_id: str
    organizer_email: str
    attendees: list[CalendarAttendee] = []
    location: str | None = None
    description: str | None = None
    video_conference: bool = False
    recurrence_rule: str | None = None
    reminders: list[dict[str, Any]] = []
    conflict_override: bool = False
    conflict_summary: list["BusySlot"] = []
    external_event_id: str | None = None
    calendar_action: str | None = None


class CalendarEventResponse(BaseModel):
    """Calendar 事件详情。"""

    id: str
    calendar_id: str
    summary: str | None = None
    description: str | None = None
    location: str | None = None
    start: CalendarEventTime | None = None
    end: CalendarEventTime | None = None
    attendees: list[CalendarAttendee] = []
    html_link: str | None = None
    raw: dict[str, Any] = {}


class ListEventsRequest(BaseModel):
    """读取未来日程请求。"""

    calendar_id: str = "primary"
    time_min: str
    time_max: str | None = None
    max_results: int = Field(default=10, ge=1, le=50)


class ListEventsResponse(BaseModel):
    """读取未来日程响应。"""

    events: list[CalendarEventResponse]


class FreebusyRequest(BaseModel):
    """Freebusy 查询请求。"""

    time_min: str
    time_max: str
    timezone: str
    calendar_ids: list[str] = ["primary"]


class BusySlot(BaseModel):
    """忙碌时间段。"""

    calendar_id: str
    start: str
    end: str


class FreebusyResponse(BaseModel):
    """Freebusy 查询响应。"""

    busy: list[BusySlot]
    conflicts: list[BusySlot]
    warnings: list[str] = []
    raw: dict[str, Any] = {}


class PrepareCalendarEventRequest(BaseModel):
    """准备日程本地草稿的请求。"""

    thread_id: str | None = None
    title: str
    start_time: str
    end_time: str | None = None
    duration_minutes: int | None = Field(default=None, ge=1, le=24 * 60)
    timezone: str
    calendar_id: str = "primary"
    organizer_email: str
    attendees: list[CalendarAttendee] = []
    location: str | None = None
    description: str | None = None
    video_conference: bool = False
    recurrence_rule: str | None = None
    conflict_override: bool = False


class CalendarArtifactResponse(BaseModel):
    """本地日程 Artifact 响应。"""

    thread_id: str
    work_item_id: str
    artifact_id: str
    version: int
    maturity: str
    conflicts: list[BusySlot]
    warnings: list[str]
    content: CalendarEventPayload
